libecole/include/ecole/scip/model.hpp

```diff
@@ -96,8 +99,23 @@ class Model {
 	 *
 	 * @see get_param_explicit, set_param_explicit to avoid any conversions.
 	 */
-	template <typename T> void set_param(std::string const& name, T value);
-	template <typename T> T get_param(std::string const& name) const;
+	// setter specialization for strings
+	void set_param(std::string const& name, std::string const& value);
+	// setter specialization for arithmetic types
+	template <typename T>
+	typename std::enable_if<std::is_arithmetic<typename std::decay<T>::type>::value, void>::
+		type
+		set_param(std::string const& name, T value);
+	// getter specialization for strings
+	template <typename T>
+	typename std::
+		enable_if<std::is_same<typename std::decay<T>::type, std::string>::value, T>::type
+		get_param(std::string const& name) const;
+	// getter specialization for arithmetic types
+	template <typename T>
+	typename std::enable_if<std::is_arithmetic<typename std::decay<T>::type>::value, T>::
+		type
+		get_param(std::string const& name) const;
```

Comment:
> Also, what is the reason for moving the file src/scip/utils.cpp?
You mean the header file, `src/scip/utils.hpp`. At some point I was using `call()` in a template function, from a hpp file, therefore I needed the function declaration in a header file. In the end I'm not having that any more, so there is no need for it in this commit. I will remove it.

However, I believe the `call()` function should not be made private. It should be available for users to call SCIP functions and let ECOLE handle exceptions. What do you think ?

Reply:
> I would prefer _not_ to change the signature here.
> IMHO the SFINAE is an implementation detail here and this is hard to read for the user.

One advantage with this new implementation is that if users throw anything else than a numeric (char, integer, floating point) or a string, then they will get an error at compile-time. This is not the case with the current implementation, the error will appear as a runtime exception.

> Think that this signature will show up in the doc

I agree on this point though. I will think of a way to have the old signature for `set_param` and `get_param`, and yet having safe type-checking at compile time.

Reply:
> However, I believe the call() function should not be made private. It should be available for users to call SCIP functions and let ECOLE handle exceptions. What do you think ?

Makes sense.

> agree on this point though. I will think of a way to have the old signature for set_param and get_param, and yet having safe type-checking at compile time.

I think the structure is already here, you just need to apply your logic to the `internal::Caster` instead of `set_param` `get_param`.
Given that it is a template method, it will _at compile time_ instantiate a `Caster`, and then the `Caster` can fail if needed.